Update developers/dogstatsd/high_throughput.md to add optimization sections (#7248)
DataDog/documentation: content/en/developers/dogstatsd/high_throughput.md

## General tips

### Use Datadog official clients

Official Dogstatsd clients are provided by Datadog for every major programming language.

> Their features are optimized for usage with Dogstatsd, please use them if one exist

Reviewer: ```suggestion
Their features are optimized for usage with Dogstatsd, please use them if one exists
```

dogstatsd_so_rcvbuf: 26214400
```

### Over UDS (Unix Domain Socket)

#### Linux

> Some Linux distributions have a kernel datagram queue size which can be low. This can be confirmed using the following command:

Author: I've added detail what it is and slight details on how Linux is using that value / working internally.

```bash
sysctl net.unix.max_dgram_qlen
```

If the value is < 512, you can increase it to 512 or more using this command:

```bash
sysctl -w net.unix.max_dgram_qlen=512
```

Add the following configuration to `/etc/sysctl.conf` to make this change permanent:

```conf
net.unix.max_dgram_qlen = 512
```

> In the same manner, the `net.core.wmem_max` could be incremented to 25MiB to improve

Author: In order to have a controlled behavior, let's ask to force the value to `dogstatsd_so_rcvbuf` as well 👍 This way, no default or unmanaged value are in the pipe.

client writing performances:

```conf
net.core.wmem_max = 26214400
```

> ## Configuration optimization guide

Reviewer: Most of this section probably shouldn't live in our public docs:

* I think that the config options mentioned here should not be considered part of the "officially-supported config options", so that we have room to break them in the future as we improve performance and perf options in the daemon. But let me know what you think of course.
* overall this seems a bit too complex for end users, and goes into lots of internal details that are very interesting to us but not really to the end user.

For now, in the public docs here, I think we should stick to general recommendations like:

* recommend to use the latest version of an official dogstatsd client (which should have best performance and by default use the right defaults for UDS and UDP datagram sizes -- if it's not the case now for some clients it will be in a couple of months)
* mention these "optimal" datagram sizes for users using non-official clients.
* if we add a setting that sets at a high level in the Agent the sizes of the buffers, only present here this single setting, and describe what the tradeoff is

Note for UDP: as the UDP packets are most of the time going through the Ethernet and IP layer, it is be better to avoid IP packets fragmentation by limiting the packet size to a value lower than a single Ethernet frame on your network. Most of the time, IPv4 networks are configured with a MTU of 1500 bytes, in this situation the packet size of sent packets should be limited to 1472.

Note for UDS: for the best performances, UDS packet should have a size of 8192 bytes.

### Limit the maximum memory usage of the Agent

The Agent tries to absorb burst of metrics sent by the Dogstatsd clients but in order to do so, it needs to use memory. Even if this is for a short amount of time and even if this memory is quickly released to the OS, a spike happens and that could be an issue in containerized environments where limit on memory usage could evict pods or containers.

> This first thing to do to not have the Dogstatsd server within the agent to reach its maximum memory usage is to avoid sending metrics in bursts in your application.

Reviewer 2: ```suggestion
Avoid sending metrics in bursts in your application - this prevents the Datadog Agent from reaching its maximum memory usage.
```

> Another thing to look at to limit the maximum memory usage is to reduce the buffering. The main buffer of the Dogstatsd server within the Agent is configurable with the `dogstatsd_queue_size` field (since Datadog Agent 6.1.0), its default value of `1024` induces an approximate maximum memory usage of 768MB.

Reviewer 2: ```suggestion
Another thing to look at to limit the maximum memory usage is to reduce the buffering. The main buffer of the DogStatsD server within the Agent is configurable with the `dogstatsd_queue_size` field (since Datadog Agent 6.1.0), its default value of `1024` induces an approximate maximum memory usage of 768MB.
```

> Note that reducing this buffer could have the caveat of increasing the number of packet drops.

Reviewer 2: ```suggestion
**Note**: reducing this buffer could increase the number of packet drops.
```

> Find below an example decreasing the max memory usage of Dogstatsd to approximately 384MB:

Reviewer 2: ```suggestion
This example decreases the max memory usage of DogStatsD to approximately 384MB:
```

```yaml
dogstatsd_queue_size: 512
```

## Client side telemetry

> Dogstatsd clients send telemetry metrics by default to the Agent. This allows you to better troubleshoot where bottleneck exists. Each metric will be tagged with the client language and the client version. These metrics will not be counted as custom metrics and will not be billed.

Reviewer 2: ```suggestion
DogStatsD clients send telemetry metrics by default to the Agent. This allows you to better troubleshoot where bottlenecks exist. Each metric is tagged with the client language and the client version. These metrics are not counted as custom metrics.
```

```conf
net.unix.max_dgram_qlen = 512
```

In the same manner, the `net.core.wmem_max` could be incremented to 4MiB to improve
client writing performances:

```conf
net.core.wmem_max = 4194304
```

Then set the Agent `dogstatsd_so_rcvbuf` configuration option to the same number in `datadog.yaml`:

```yaml
dogstatsd_so_rcvbuf: 4194304
```

Please note that setting these values will need a reboot of the host.

### Ensure proper packet sizes

Sending packets with an adequate size to Dogstatsd will avoid extra work to the server.

If the packets sent are too small, Dogstatsd will have to pack several together in order
to process them in batches later in the pipeline. The official Dogstatsd clients are
capable of grouping metrics to have the best ratio of the number of metrics per packet.

If the clients properly send packets having a size close to `dogstatsd_buffer_size`, Dogstatsd
will perform better. Be careful: the packets must not be larger than this size, otherwise, the
Dogstatsd server won't be able to load them completely in the buffer and some of metrics will
be malformed. Please use the corresponding configuration field in your Dogstatsd clients.

> Note for UDP: as the UDP packets are most of the time going through the Ethernet and IP layer,
> they inherently have a size limited to 1472.

Reviewer 3: 1472 is the "usual limit" (IPv4 header with no option/1500 bytes Ethernet MTU) and it may be larger (bigger MTU) or smaller (due to IPv6 larger header).
```suggestion
Note for UDP: as the UDP packets are most of the time going through the Ethernet and IP layer, fragmentation should be avoided even if supported. Packet size should be limited to fit at most a single Ethernet frame and prevent IP fragmentation.
```
I'm not sure the suggested wording is understandable, but I think we should definitely mentioned that IP fragmentation should be avoided.

Author: You're right I'll add more precision here

Author: I've addressed your feedback, please re-open if you think that's not clear enough!

Note for UDS: for the best performances, UDS packet should have a size of 8192 bytes.

### Limit the maximum memory usage

Dogstatsd tries to absorb burst of metrics sent by the clients but in order to do so,
it needs to use memory. Even if this is for a short amount of time and even if this
memory is quickly released to the OS, a spike happens and that could be an issue in
containerized environments where limit on memory usage could evict pods or containers.

This first thing to do to not have Dogstatsd reach its maximum memory usage is
to avoid sending metrics in bursts in your application.

Another thing to look at to limit the maximum memory usage is to reduce the buffering.

> The main buffer of Dogstatsd is configurable with the `dogstatsd_queue_size` field, its default value of `1024` induce an approximate maximum memory usage of 768MB. Note that reducing this

Reviewer 4: `induce` => `induces`?

dogstatsd_so_rcvbuf: 26214400
```

### Over UDS (Unix Domain Socket)

#### Linux

For UDS sockets, Linux is internally buffering datagrams in a queue if the
reader is slower than the writer. The size of this queue represents the maximum number
of datagrams that Linux will buffer per socket. This value can be queried with the following command:

```bash
sysctl net.unix.max_dgram_qlen
```

If the value is < 512, you can increase it to 512 or more using this command:

```bash
sysctl -w net.unix.max_dgram_qlen=512
```

Add the following configuration to `/etc/sysctl.conf` to make this change permanent:

```conf
net.unix.max_dgram_qlen = 512
```

In the same manner, the `net.core.wmem_max` could be incremented to 4MiB to improve
client writing performances:

```conf
net.core.wmem_max = 4194304
```

Then set the Agent `dogstatsd_so_rcvbuf` configuration option to the same number in `datadog.yaml`:

```yaml
dogstatsd_so_rcvbuf: 4194304
```

> Please note that setting these values will need a reboot of the host.

Reviewer: is this true for all the options mentioned here? 🤔

Pretty sure at least `net.core.wmem_max` is taken into account at runtime without a reboot?

Can be addressed in a follow-up PR in any case

Author: No you're right, I've first added this here to be cautious but I read yesterday in kubernertes doc that net sysctls doesn't need a reboot! I'll remove that.

### Ensure proper packet sizes

> Sending packets with an adequate size to Dogstatsd will avoid extra work to the server.

Reviewer: let's keep naming consistent:

```suggestion
Sending packets with an adequate size to Dogstatsd will avoid extra work to the Agent.
```

If the packets sent are too small, Dogstatsd will have to pack several together in order
to process them in batches later in the pipeline. The official Dogstatsd clients are

> capable of grouping metrics to have the best ratio of the number of metrics per packet.

Reviewer: Since this section is basically only relevant for readers who do not use the official clients, I think it's worth highlighting this first. Something like this as the second sentence of this section could work for example:

```
The latest versions of the official Dogstatsd clients send packets with a size optimized for performance. If you're using them, skip this section, otherwise read on.
```

(not super important, can be addressed separately)

If the clients properly send packets having a size close to `dogstatsd_buffer_size`, Dogstatsd
will perform better. Be careful: the packets must not be larger than this size, otherwise, the
Dogstatsd server won't be able to load them completely in the buffer and some of metrics will
be malformed. Please use the corresponding configuration field in your Dogstatsd clients.

Note for UDP: as the UDP packets are most of the time going through the Ethernet and IP layer,
it is be better to avoid IP packets fragmentation by limiting the packet size to a value lower
than a single Ethernet frame on your network. Most of the time, IPv4 networks are configured with
a MTU of 1500 bytes, in this situation the packet size of sent packets should be limited to 1472.

Note for UDS: for the best performances, UDS packet should have a size of 8192 bytes.

> ### Limit the maximum memory usage

Reviewer: ```suggestion
### Limit the maximum memory usage of the Agent
```

> Dogstatsd tries to absorb burst of metrics sent by the clients but in order to do so,

Reviewer: AFAIK the terminology we use in our public docs is:

* `Dogstatsd`: the protocol
* the `Agent` (or `Dogstatsd-standalone binary`): the dogstatsd server
* `Dogstatsd clients`: (obviously) the clients :)

So in this section, I think we should replace most (all?) instances of `Dogstatsd` with `Agent`. If we're talking about the memory usage of the Dogstatsd bits of the Agent, I guess we can write `the Dogstatsd server within the Agent`, or something of that kind, WDYT?

Author: 👍

it needs to use memory. Even if this is for a short amount of time and even if this
memory is quickly released to the OS, a spike happens and that could be an issue in
containerized environments where limit on memory usage could evict pods or containers.

This first thing to do to not have Dogstatsd reach its maximum memory usage is
to avoid sending metrics in bursts in your application.

Another thing to look at to limit the maximum memory usage is to reduce the buffering.

> The main buffer of Dogstatsd is configurable with the `dogstatsd_queue_size` field, its default value of `1024` induces an approximate maximum memory usage of 768MB. Note that reducing this

Reviewer: let's mention since which agent version this config option exists
